Userland/Utilities/mktemp.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2021, Idan Horowitz <idan.horowitz@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/LexicalPath.h>
+#include <LibCore/ArgsParser.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+constexpr const char* default_template = "tmp.XXXXXXXXXX";
+
+static const char* generate_random_filename(const char* pattern)
+{
+    char* new_filename = strdup(pattern);
+    int pattern_length = strlen(pattern);
+
+    static constexpr char random_characters[] = "abcdefghijklmnopqrstuvwxyz0123456789";
+    for (auto i = pattern_length - 1; i >= 0; --i) {
+        if (pattern[i] != 'X')
+            break;
+        new_filename[i] = random_characters[(arc4random() % (sizeof(random_characters) - 1))];
+    }
+
+    return new_filename;
+}
+
+static const char* make_temp(const char* pattern, bool directory, bool dry_run)
```

Comment:
Is there anything special about the `mktemp(1)` utility that means you can't use `mktemp(3)` and `mkdtemp` from stdlib.h? I guess maybe the dry run option?

Reply:
> 
> 
> Is there anything special about the `mktemp(1)` utility that means you can't use `mktemp(3)` and `mkdtemp` from stdlib.h? I guess maybe the dry run option?

`mktemp` can be used to perform a dry run. It does not create the associated file.

Reply:
mktemp(3) only allows templates with exactly 6 Xs (6 random characters), while mktemp(1) is specified to allow any amount larger than 3, but yes the dry run option as well